jkq/ddsim/pathqasmsimulator.py

```diff
@@ -0,0 +1,271 @@
+"""Backend for DDSIM Task-Based Simulator."""
+
+import logging
+import os
+
+import time
+import uuid
+import warnings
+from typing import Union, List
+
+from qiskit.providers import BackendV1, Options
+from qiskit import QiskitError, QuantumCircuit
+from qiskit.providers.models import BackendConfiguration, BackendStatus
+from qiskit.qobj import QasmQobjExperiment, Qobj, QasmQobj, PulseQobj
+from qiskit.result import Result
+from qiskit.compiler import assemble
+
+from .jkqjob import JKQJob
+from .jkqerror import JKQSimulatorError
+from jkq import ddsim
+
+logger = logging.getLogger(__name__)
+
+
+def read_tensor_network_file(filename):
+    import pandas as pd
+    import numpy as np
+    import quimb.tensor as qtn
+
+    df = pd.read_json(filename)
+    tensors = []
+
+    for i in range(len(df['tensors'])):
+        tens_data = [complex(real, imag) for [real, imag] in df['tensors'][i][3]]
+        tens_shape = df['tensors'][i][2]
+        data = np.array(tens_data).reshape(tens_shape)
+        inds = df['tensors'][i][1]
+        tags = df['tensors'][i][0]
+        tensors.append(qtn.Tensor(data, inds, tags, left_inds=inds[:len(inds) // 2]))
+    return tensors
+
+
+def create_tensor_network(qc):
+    import quimb.tensor as qtn
+    import sparse
+    from jkq.ddsim import dump_tensor_network
+
+    if isinstance(qc, QuantumCircuit):
+        filename = qc.name + '_' + str(qc.num_qubits) + ".tensor"
+        nqubits = qc.num_qubits
+    else:
+        filename = 'tensor.tensor'
+        nqubits = qc.header.n_qubits
+
+    dump_tensor_network(qc, filename)
+    tensors = read_tensor_network_file(filename)
+    os.remove(filename)
+
+    # add the zero state tensor |0...0> at the beginning
+    shape = [2] * nqubits
+    data = sparse.COO(coords=[[0]] * nqubits, data=1, shape=shape, sorted=True, has_duplicates=False)
+    inds = ['q' + str(i) + '_0' for i in range(nqubits)]
+    tags = ['Q' + str(i) for i in range(nqubits)]
+    tensors.insert(0, qtn.Tensor(data=data, inds=inds, tags=tags))
+
+    # using the following lines instead would allow much greater flexibility,
+    # but is not supported for DD-based simulation at the moment
+
+    # add the zero state tensor |0>...|0> at the beginning
+    # shape = [2]
+    # data = np.zeros(2)
+    # data[0] = 1.0
+    # for i in range(qc.num_qubits):
+    #     inds = ['q'+str(i)+'_0']
+    #     tags = ['Q'+str(i)]
+    #     tensors.insert(0, qtn.Tensor(data=data.reshape(shape), inds=inds, tags=tags))
+
+    return qtn.TensorNetwork(tensors)
+
+
+def get_simulation_path(qc, max_time: int = 60, max_repeats: int = 1024, dump_path: bool = True,
+                        plot_ring: bool = False):
+    import cotengra as ctg
+    from opt_einsum.paths import linear_to_ssa
+
+    tn = create_tensor_network(qc)
+
+    opt = ctg.HyperOptimizer(max_time=max_time,
+                             max_repeats=max_repeats,
+                             progbar=True,
+                             parallel=4,
```

Comment:
I assume `parallel` should not have a fixed value here, but should rather be configurable as the other settings.